README.md

# PrometheeTools

<!-- badges: start -->
[![R-CMD-check](https://github.com/ifelipebj/PrometheeTools/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/ifelipebj/PrometheeTools/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->


The goal of PrometheeTools is to provide PROMETHEE (Preference Ranking Organisation
METHod for Enrichment of Evaluations), a multicriteria decision-making method, to
handle a large number of alternatives to obtain partial and complete rankings.
Moreover, the package provides the GLNF (Global Local Net Flow) sorting algorithm to
classify alternatives into ordered categories and the SILS function to measure the
classification quality based on sorting methods. 

## Installation
You can install the released version of PrometheeTools from
[CRAN](https://CRAN.R-project.org/package=PrometheeTools) with:

```r
install.packages("PrometheeTools")
```

You can install the development version of PrometheeTools from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("ifelipebj/PrometheeTools")
```
## Functions

#### PROMETHEE Outranking Method

PROMETHEE is a multicriteria method that quantifies preference relationships and
obtains the positive, negative and net flows of the alternatives, generating
rankings that reflect the decision-maker's preferences. This function applies
PROMETHEE I (partial ranking) and PROMETHEE II (complete ranking). This function 
can handle  large number of alternatives.

Usage:

``` r
PROMETHEEII(matrix_evaluation, data_criteria)
```
#### Global and Local Searches for Net Flows to Sort

This function applies the GLNF Sorting (Global Local Net Flow Sorting)
algorithm to classify the alternatives into ordered groups according to the
decision-maker's preferences in multiple criteria context. GLNF sorting is
based on PROMETHEE net flows and a set of limiting profiles. This algorithm
starts from a global classification (global search) that is enhanced by two
local searches, intra-categorical and inter-categorical.

Usage:

``` r
GLNF(matrix_evaluation, data_criteria)
```
#### Quality Index of Silhouette for Sorting

This function computes a quality index for `SILS` (Silhouette for Sorting),
which relies on PROMETHEE II net flows to assess the classifications produced
by PROMETHEE-based ordered sorting methods.

Usage:

``` r
SILS(matrix_evaluation, data_criteria, k, SILS_plot = TRUE)
```
## Evaluation matrix template

The evaluation matrix for the `evaluation_matrix` is formed by a set of alternatives
A = (a1,a2,...,a_i,...,a_n) and a set of criteria G = (g1,g2,...,g_j,...,g_m), it must
be structured as follows:

| Alternative   | Criterion 1 | Criterion 2 | ... | Criterion j | ... | Criterion m |
|---------------|-------------|-------------|-----|-------------|-----|-------------|
| Alternative 1 |             |             |     |             |     |             |
| Alternative 2 |             |             |     |             |     |             |
| ...           |             |             |     |             |     |             |
| Alternative i |             |             |     |             |     |             |
| ...           |             |             |     |             |     |             |
| Alternative n |             |             |     |             |     |             |


Where:
- `Alternative` are the names or identifiers of the alternatives being 
evaluated.
- `Criterion 1`, `Criterion 2`, etc. are placeholders for the specific names of 
your criteria.
- Each cell in the matrix represents the value of the corresponding criterion 
for a specific alternative.

The incorporation of limiting profiles that define the category in the
`matrix_evaluation` is a crucial step to apply the `GLNF` and `SILS` 
functions. These profiles are denoted using the syntax "r" followed by the 
profile number. For sorting into `k` groups, there should defined `k+1` 
limiting profiles with R=(r1,r2,...,r`k+1`) being the set of limiting profiles.
"r1" and "r2" represent the limiting profiles for the first group (most
preferred), and "r`k`" and "r`k+1`" denote the profiles for the `k`-th group 
(least preferred). Finally, the set A and R are added into a new set of 
alternatives Z = (a1,a2,...,a_n,r1,r2,...,r`k+1`).

To implement `SILS`,is necessary to add a new column `Category` at the end of 
the `evaluation_matrix` with the classification in (the alternatives assignment to
categories) to be evaluated. Write the classification in number format 
(e.g.: 1,2,3,...,`k`).

Replace the placeholders with the actual names of your criteria and provide
appropriate values for each alternative-criterion combination.


## Criteria parameters table template

The criteria parameters table for the `data_criteria` should be structured as follows:

| Parameter              | Criterion 1 | Criterion 2 | ... | Criterion j | ... | Criterion m |
|------------------------|-------------|-------------|-----|-------------|-----|-------------|
| Function Type          |             |             |     |             |     |             |
| Indifference Threshold |             |             |     |             |     |             |
| Preference Threshold   |             |             |     |             |     |             |
| Objective              |             |             |     |             |     |             |
| Weight                 |             |             |     |             |     |             |

Where:
- `Criterion 1`, `Criterion 2`, etc. are placeholders for the specific names of 
your criteria. 
- `Function Type` specifies the type of function for the criterion ("linear",
"v-shape", "usual","u-shape", "level" and "gaussian").
- `Indifference Threshold` is the indifference threshold for the criterion.
- `Preference Threshold` is the preference threshold for the criterion.
- `Objective` specifies whether the criterion's objective is to maximize ("max") 
or minimize ("min").
- `Weight` represents the weight of the criterion in the analysis. The sum of the
weights of all criteria must be equal to 1.

The preference and indifference thresholds depend on the type of function
selected. The preference threshold is non-zero for all functions except for
"usual" and "u-shaped". The indifference threshold is non-zero for "linear",
"level" and "u-shaped" functions.

Replace the placeholders with the actual names of your criteria and provide
appropriate values for each parameter.


## Example

This is an example which shows you how to solve a ranking and sorting problem:

``` r
# Load the PrometheeTools library
library(PrometheeTools)
# Example code corresponds to 30 customers to be evaluated with five 
# criteria for partial ranking (PROMETHEE I), complete ranking (PROMETHEE II) and 
# sorting into four groups ordered by decision-maker preferences (GLNF sorting).
# Finally the silhouette quality index for sorting (`SILS`) is calculated.

# The evaluation matrix is defined, and five limiting profiles are included 
# (r1,r2,r3,r4,r5) to define four categories.

#############

matrix_evaluation <- data.frame (
 
Alternative = c(1, 2, 3, 4, 5, 6, 7, 8, 9, 10,
                11, 12, 13, 14, 15, 16, 17, 18, 19, 20,
                21, 22, 23, 24, 25, 26, 27, 28, 29, 30,
                "r1", "r2", "r3", "r4", "r5"),
Monetary = c(21.52, 68.09, 184.94, 237.62, 14.29, 12.78, 91.53, 11.39, 264.79, 12.74,
             274.41, 3.75, 47.92, 34.5, 45.89, 39.92, 31.18, 273.23, 16.39, 3.91,
             20.09, 6.52, 26.62, 28.47, 7.57, 69.2, 420.95, 12.01, 85.88, 8.78,
             6816.80, 120, 40, 20, 0),
Recency = c(0, 0, 0, 0, 3, 5, 0, 6, 0, 3,
            1, 0, 1, 0, 0, 0, 0, 0, 2, 1,
            0, 0, 0, 0, 5, 1, 0, 0, 1, 4,
            0, 1, 7, 8, 12),
Frequency = c(7, 5, 12, 12, 1, 3, 9, 2, 12, 4,
              11, 3, 10, 10, 11, 11, 12, 12, 7, 1,
             5, 2, 9, 11, 4, 10, 12, 3, 10, 2,
             12, 10, 8, 4, 1),
Financial_score = c(66, 58, 83, 68, 68, 69, 77, 55, 77, 53,
                    78, 35, 84, 75, 71, 64, 56, 55, 52, 30,
                    66, 50, 65, 53, 54, 82, 68, 53, 62, 43,
                     100, 80, 75, 65, 0),
                     
  Length = c(4, 3, 3, 2, 2, 2, 2, 3, 2, 4,
            3, 3, 1, 1, 2, 5, 4, 2, 2, 5,
            4, 5, 1, 4, 2, 1, 5, 1, 1, 2,
            5, 4, 3, 2, 1))
  
# The matrix with the criteria parameters is defined
  
  data_criteria <- data.frame(
  Parameter = c("Function Type", "Indifference Threshold",
               "Preference Threshold","Objetive", "Weight"),
  Frequency = c("linear", 0, 3, "max", 0.2),
  Monetary = c("linear", 30.00, 120, "max", 0.4),
  Recency = c("usual", 0.00, 0.00, "min", 0.1),
  Financial_score = c("linear", 0.00, 10, "max", 0.2),
  Length = c("usual", 0.00, 0.00, "max", 0.1))

############# 

# The `PROMETHEEII` function is applied
RS <- PROMETHEEII(matrix_evaluation, data_criteria)
# Positive flows, negative flows and net flows
RS$NF
# Net flows by criteria
RS$NFC

#############

# The `GLNF` function is applied
RS <- GLNF(matrix_evaluation, data_criteria)
# Final classification of the 30 customers into four ordered groups
RS$Class
# It is possible to check the details of global and local searches.
# Show global classification results
RS$Global
# Show local classification results
RS$Local1
RS$Local2

#############

# Add the categories calculated with GLNF (or another method) to the last column 
# of `matrix_evaluation`. The limiting profiles have not been assigned to any
# group.

matrix_evaluation$Category = c(3, 3, 1, 1, 4, 3, 2, 4, 1, 3,
                                 1, 4, 2, 2, 2, 2, 2, 1, 3, 4,
                                 3, 3, 3, 2, 4, 2, 1, 4, 3, 4,
                                 NA, NA, NA, NA, NA)

# The number of categories or groups to be evaluated is defined.

k <- 4

# The function `SILS` is applied, where `SILS_plot = TRUE` to plot the results 
# of the silhouettes.

RS <- SILS(matrix_evaluation, data_criteria, k, SILS_plot = TRUE)
#Print silhouette values
print(RS)

```



NEWS.md
# PrometheeTools (development version)

* Initial CRAN submission.


R/SILS.R
#' Quality Index of Silhouette for Sorting
#'
#' This function computes the quality index for `SILS` (Silhouette for Sorting),
#' which relies on PROMETHEE II net flows to assess the classifications generated
#' by PROMETHEE-based ordered sorting methods.
#' @param matrix_evaluation The matrix includes the values for all alternatives
#' and limiting profiles are rows and columns correspond to the evaluation
#' criteria. The last column indicates the alternative classification.
#' @param data_criteria Matrix with the parameter information (rows) for each
#' criterion (columns). The rows of parameters are in the following order:
#' Function Type, Indifference Threshold, Preference Threshold, Objective and
#' Weight.
#' @param k The number of categories to be evaluated.
#' @param SILS_plot Boolean value indicating whether to generate a stacked bar
#' chart representing the SILS values.
#' @references
#' Barrera, F., Segura, M., & Maroto, C. (2023) Online. Multicriteria sorting method
#' based on global and local search for supplier segmentation. International
#' Transactions in Operational Research. DOI:10.1111/itor.13288
#' @details
#' - The categories corresponding to the classifications to be assessed should
#' be indicated in the last column of `matrix_evaluation` in ordinal numbers,
#' where 1 is the most preferred group.
#' - Enter the same criteria parameters and limiting profiles that you used to
#' obtain the classifications with PROMETHEE II.
#' - The Limiting Profiles used must be presented as rows in the
#' matrix_evaluation. The name must start with the letter "r" followed by the
#' profile number (e.g., "r1", "r2"). For `k` categories, there
#' should define `k + 1` limiting profiles.
#' - The types of preference function are as follows: "linear", "v-shape",
#' "usual", "u-shape", "level" and "gaussian".
#' - The preference and indifference thresholds depend on the type of function
#' selected. The preference threshold requires definition (is non-zero) for all
#' functions except for "usual" and "u-shaped". The indifference threshold is
#' non-zero for "linear", "level" and "u-shaped" functions.
#' - In the objective write "max" to maximize or "min" to minimize.
#' - The sum of the weights of all criteria must be equal to 1.
#' @return
#' - A data frame with the SILS values for each alternative according to the
#' classification entered.
#'
#' - Stacked bar chart representing the SILS values for each alternative with
#' horizontal control limits.
#' @export
#' @seealso \code{\link{PROMETHEEII}}
#' @importFrom ggplot2 ggplot aes geom_bar scale_fill_manual theme_minimal
#' labs geom_hline scale_x_discrete scale_y_continuous theme element_text margin
#' @examples
#' k <- 4
#' matrix_evaluation <- data.frame (
#'
#'Alternative = c(1, 2, 3, 4, 5, 6, 7, 8, 9, 10,
#'                  11, 12, 13, 14, 15, 16, 17, 18, 19, 20,
#'                  21, 22, 23, 24, 25, 26, 27, 28, 29, 30,
#'                 "r1", "r2", "r3", "r4", "r5"),
#'Monetary = c(21.52, 68.09, 184.94, 237.62, 14.29, 12.78, 91.53, 11.39, 264.79, 12.74,
#'             274.41, 3.75, 47.92, 34.5, 45.89, 39.92, 31.18, 273.23, 16.39, 3.91,
#'             20.09, 6.52, 26.62, 28.47, 7.57, 69.2, 420.95, 12.01, 85.88, 8.78,
#'             6816.80, 120, 40, 20, 0),
#'Recency = c(0, 0, 0, 0, 3, 5, 0, 6, 0, 3,
#'            1, 0, 1, 0, 0, 0, 0, 0, 2, 1,
#'            0, 0, 0, 0, 5, 1, 0, 0, 1, 4,
#'            0, 1, 7, 8, 12),
#'Frequency = c(7, 5, 12, 12, 1, 3, 9, 2, 12, 4,
#'              11, 3, 10, 10, 11, 11, 12, 12, 7, 1,
#'              5, 2, 9, 11, 4, 10, 12, 3, 10, 2,
#'              12, 10, 8, 4, 1),
#'Financial_score = c(66, 58, 83, 68, 68, 69, 77, 55, 77, 53,
#'                    78, 35, 84, 75, 71, 64, 56, 55, 52, 30,
#'                    66, 50, 65, 53, 54, 82, 68, 53, 62, 43,
#'                    100, 80, 75, 65, 0),
#'Length = c(4, 3, 3, 2, 2, 2, 2, 3, 2, 4,
#'           3, 3, 1, 1, 2, 5, 4, 2, 2, 5,
#'           4, 5, 1, 4, 2, 1, 5, 1, 1, 2,
#'           5, 4, 3, 2, 1),
#'Category = c(3, 3, 1, 1, 4, 3, 2, 4, 1, 3,
#'               1, 4, 2, 2, 2, 2, 2, 1, 3, 4,
#'               3, 3, 3, 2, 4, 2, 1, 4, 3, 4,
#'               NA, NA, NA, NA, NA))
#'data_criteria <- data.frame(
#'  Parameter = c("Function Type", "Indifference Threshold",
#'                "Preference Threshold","Objetive", "Weight"),
#'  Frequency = c("linear", 0, 3, "max", 0.2),
#'  Monetary = c("linear", 30.00, 120, "max", 0.4),
#'  Recency = c("usual", 0.00, 0.00, "min", 0.1),
#'  Financial_score = c("linear", 0.00, 10, "max", 0.2),
#'  Length = c("usual", 0.00, 0.00, "max", 0.1))
#'RS <- SILS(matrix_evaluation, data_criteria, k, SILS_plot = TRUE)
#'print(RS)
SILS <- function(matrix_evaluation, data_criteria, k, SILS_plot = FALSE) {
  colnames(matrix_evaluation)[1] <- "Alternative"
  colnames(matrix_evaluation)[ncol(matrix_evaluation)] <- "Category"
  Category <- matrix_evaluation$Category[!is.na(matrix_evaluation$Category)]
  matrix_evaluation <- matrix_evaluation[, -which(names(matrix_evaluation) == "Category")]
  # Calculate PROMETHEE
  RS <- PROMETHEEII(matrix_evaluation, data_criteria)
  data <- as.data.frame(RS[[1]])
  data$Phi <- as.numeric(data$Phi)
  # Prepare data for SILS
  lim_prof <- data[grepl("^r\\d+", data$Alternative), ]
  lim_prof <- lim_prof[order(as.numeric(gsub("^r", "", lim_prof$Alternative))), ]
  r <- lim_prof$Phi
  data <- data[!grepl("^r\\d+", data$Alternative), ]
  data <- cbind(data, Category = Category)
  data <- data[order(data$Category, -data$Phi), ]
  Category <- data$Category
  phi <- data$Phi
  Alternative <- data$Alternative
  # Define variable d_t
  d_t <- 2 / ((k + 1) * 2)
  # Define Category centroids
  centroids <- (r[-length(r)] + r[-1]) / 2
  # Create vectors for dissimilarities
  u <- rep(0, length(phi))
  l <- rep(0, length(phi))
  h <- rep(0, length(phi))
  SILS <- rep(0, length(phi))
  # Current group dissimilarity (u(i))
  for (i in seq_along(phi)){
    cat_i <- Category[i]
    phi_i <- phi[i]
    num_alt_cat_i <- sum(Category == cat_i)
    num_alt_cat_i
    dissim_i <- sum(abs(phi[Category == cat_i] - phi_i))
    dissim_i
    dissicentro_k <- abs((phi_i - centroids[cat_i]))
    u[i] <- (dissim_i + dissicentro_k) / (num_alt_cat_i)
  }
  # Lower group dissimilarity (l(i))
  for (i in seq_along(phi)) {
    cat_i <- Category[i]
    phi_i <- phi[i]
    num_alt_cat_i <- sum(Category == cat_i + 1)
    if (cat_i == k) {
      l[i] <- d_t
    } else {
      dissim_i <- sum((phi_i - phi[Category == cat_i + 1]))
      dissicentro_k <- ((phi_i - centroids[cat_i + 1]))
      l[i] <- (dissim_i + dissicentro_k) / (num_alt_cat_i + 1)
    }
  }
  # Higger group dissimilarity (h(i))
  for (i in seq_along(phi)){
    cat_i <- Category[i]
    phi_i <- phi[i]
    num_alt_cat_i <- sum(Category == cat_i - 1)
    if (cat_i == 1) {
      h[i] <- d_t
    } else {
      dissim_i <- sum((phi[Category == cat_i - 1] - phi_i))
      dissicentro_k <- ((centroids[cat_i - 1] - phi_i))
      h[i] <- (dissim_i + dissicentro_k) / (num_alt_cat_i + 1)
    }
  }
  # Silhouette sorting calculation
  for (i in seq_along(phi)){
    SILS[i] <- (l[i] - u[i]) / max(l[i], u[i]) - (h[i] - u[i]) / max(h[i], u[i])
  }
  RS <- data.frame(Alternative, Category, SILS)
  # Plot for silhouettes
  if (SILS_plot) {
    ggplot2::ggplot()
    SILS_ggplot <- ggplot(RS, aes(x = Alternative, y = SILS, fill = ifelse(SILS > 0, "", ""))) +
      geom_bar(stat = "identity", width = 0.7, color = "black") +
      scale_fill_manual(values = c("Positivo" = "#ADD8E6", "Negativo" = "#FFC0CB")) +
      theme_minimal() +
      labs(x = "Alternative", y = "SILS  Value", size = 30, family = "Times New Roman") +
      geom_hline(yintercept = 0, color = "black", linetype = "solid", linewidth = 1) +
      geom_hline(yintercept = 1, color = "gray", linetype = "dashed", linewidth = 1) +
      geom_hline(yintercept = -1, color = "gray", linetype = "dashed", linewidth = 1) +
      geom_hline(yintercept = 1.5, color = "red", linetype = "dashed", linewidth = 1) +
      geom_hline(yintercept = -1.5, color = "red", linetype = "dashed", linewidth = 1) +
      scale_x_discrete(breaks = seq_along(SILS)) +
      scale_y_continuous(limits = c(-2, 2), breaks = seq(-2, 2, by = 0.5)) +
      theme(axis.text.y = element_text(hjust = 1, margin = margin(r = 10)),
            axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
    print(SILS_ggplot)
  }
  return(RS)
}


R/GLNF.R
#'Global and Local Searches for Net Flows to Sort
#'
#'This function applies the GLNF Sorting (Global Local Net Flow Sorting)
#'algorithm to classify the alternatives into ordered groups according to the
#'decision-maker's preferences in multiple criteria context. GLNF sorting is
#'based on PROMETHEE net flows and a set of limiting profiles. This algorithm
#'starts from a global classification (global search) that is enhanced by two
#'local searches, intra-categorical and inter-categorical.
#' @usage
#' GLNF(matrix_evaluation, data_criteria)
#' @param matrix_evaluation The matrix includes the values for all alternatives.
#' The alternatives and limiting profiles are row and columns correspond to
#' the evaluation criteria.
#' @param data_criteria Matrix with the parameter information (rows) for each
#' criterion (columns). The rows of parameters are in the following order:
#' Function Type, Indifference Threshold, Preference Threshold, Objective and
#' Weight.
#' @references
#' Barrera, F., Segura, M., & Maroto, C. (2023) Online. Multicriteria sorting method
#' based on global and local search for supplier segmentation. International
#' Transactions in Operational Research. DOI:10.1111/itor.13288
#' @return
#' -`Global` Matrix with the results of the global search where positive,
#' negative and net flow, and its preclassification are defined for each
#' alternative.
#'
#' -`Local1` Matrices with the results of the first local search. PROMETHEE is
#' applied to each group obtained in the global search. The alternatives are
#' divided according to their positive or negative sign from the net flows
#' obtained from PROMETHEE.
#'
#' -`Local2` Matrices with the results of the second local search, where the
#' alternatives are divided according to their sign from net flows are obtained
#' after applying PROMETHEE between each pair of neighbour categories.
#'
#' -`Class` Final classification of the alternatives results.
#' @export
#' @seealso \code{\link{PROMETHEEII}}
#' @details
#' - The Limiting Profiles should be presented as rows in the matrix_evaluation.
#' The name must start with the letter "r" followed by the profile number
#' (e.g., "r1", "r2").
#' - For k categories, there should be k + 1 limiting profiles. To create k
#' groups the set of limit profiles are defined, where r1 is preferred to
#' r2,...,preferred to r(k+1).
#' - The types of preference function are as follows: "linear", "v-shape",
#' "usual", "u-shape", "level" and "gaussian".
#' - The preference and indifference thresholds depend on the type of function
#' selected. The preference threshold requires definition (is non-zero) for all
#' functions except for "usual" and "u-shaped". The indifference threshold is
#' non-zero for "linear", "level" and "u-shaped" functions.
#' - In the objective write "max" to maximize or "min" to minimize.
#' - The sum of the weights of all criteria must be equal to 1.
#' @examples
#'matrix_evaluation <- data.frame (
#'
#'Alternative = c(1, 2, 3, 4, 5, 6, 7, 8, 9, 10,
#'                 11, 12, 13, 14, 15, 16, 17, 18, 19, 20,
#'                 21, 22, 23, 24, 25, 26, 27, 28, 29, 30,
#'                 "r1", "r2", "r3", "r4", "r5"),
#'Monetary = c(21.52, 68.09, 184.94, 237.62, 14.29, 12.78, 91.53, 11.39, 264.79, 12.74,
#'             274.41, 3.75, 47.92, 34.5, 45.89, 39.92, 31.18, 273.23, 16.39, 3.91,
#'             20.09, 6.52, 26.62, 28.47, 7.57, 69.2, 420.95, 12.01, 85.88, 8.78,
#'             6816.80, 120, 40, 20, 0),
#'Recency = c(0, 0, 0, 0, 3, 5, 0, 6, 0, 3,
#'            1, 0, 1, 0, 0, 0, 0, 0, 2, 1,
#'            0, 0, 0, 0, 5, 1, 0, 0, 1, 4,
#'            0, 1, 7, 8, 12),
#'Frequency = c(7, 5, 12, 12, 1, 3, 9, 2, 12, 4,
#'              11, 3, 10, 10, 11, 11, 12, 12, 7, 1,
#'              5, 2, 9, 11, 4, 10, 12, 3, 10, 2,
#'              12, 10, 8, 4, 1),
#'Financial_score = c(66, 58, 83, 68, 68, 69, 77, 55, 77, 53,
#'                    78, 35, 84, 75, 71, 64, 56, 55, 52, 30,
#'                    66, 50, 65, 53, 54, 82, 68, 53, 62, 43,
#'                    100, 80, 75, 65, 0),
#'Length = c(4, 3, 3, 2, 2, 2, 2, 3, 2, 4,
#'           3, 3, 1, 1, 2, 5, 4, 2, 2, 5,
#'           4, 5, 1, 4, 2, 1, 5, 1, 1, 2,
#'           5, 4, 3, 2, 1))
#'data_criteria <- data.frame(
#'Parameter = c("Function Type", "Indifference Threshold",
#'              "Preference Threshold","Objetive", "Weight"),
#'Frequency = c("linear", 0, 3, "max", 0.2),
#'Monetary = c("linear", 30.00, 120, "max", 0.4),
#'Recency = c("usual", 0.00, 0.00, "min", 0.1),
#'Financial_score = c("linear", 0.00, 10, "max", 0.2),
#'Length = c("usual", 0.00, 0.00, "max", 0.1))
#'RS <- GLNF(matrix_evaluation, data_criteria)
#'RS$Class
#'RS$Global
#'RS$Local1
#'RS$Local2
GLNF <- function(matrix_evaluation, data_criteria) {
  colnames(matrix_evaluation)[1] <- "Alternative"
  # Step2. Global search
  RS <- PROMETHEEII(matrix_evaluation, data_criteria)
  data <- as.data.frame(RS[[1]])
  data$Phi <- as.numeric(data$Phi)
  lim_prof <- data[grepl("^r\\d+", data$Alternative), ]
  lim_prof <- lim_prof[order(as.numeric(gsub("^r", "", lim_prof$Alternative))), ]
  num_categories <- nrow(lim_prof) - 1
  phi_prof <- lim_prof$Phi
  data <- data[!grepl("^r\\d+", data$Alternative), ]
  data$Category <- NA
  # Assigning categories according to Phi
  for (i in 1:(num_categories)) {
    lim_sup <- phi_prof[i]
    lim_inf <- phi_prof[i + 1]
    if (i == 1 && any(data$Phi == lim_sup)) {
      data$Category[data$Phi == lim_sup] <- i
    }
    data$Category[data$Phi >= lim_inf & data$Phi < lim_sup] <- i
  }
  Global_Search <- data
  lim_prof$Category <- NA
  Global_Search <- rbind(Global_Search, lim_prof)
  # Creating subarrays for each group according to the global search
  Global_Search_categories <- list()
  for (i in 1:(num_categories)){
    Alt_category <- data[data$Category == i, ]
    ids_category <- Alt_category$Alternative
    lim_prof_category <- lim_prof[grepl(paste0("^r", i, "|^r", i + 1), lim_prof$Alternative), ]
    ids_category <- union(ids_category, lim_prof_category$Alternative)
    category_matrix <- matrix_evaluation[matrix_evaluation$Alternative %in% ids_category, ]
    Global_Search_categories[[paste0("C", i)]] <- category_matrix
  }
  # Step 3. Local search 1
  Local1_categories <- list()
  for (i in seq_along(Global_Search_categories)) {
    matrix_name <- names(Global_Search_categories[i])
    ME <- matrix_evaluation
    matrix_evaluation <- Global_Search_categories[[i]]
    # PROMETHEE
    RS <- PROMETHEEII(matrix_evaluation, data_criteria)
    data <- as.data.frame(RS[[1]])
    data$Phi <- as.numeric(data$Phi)
    data <- data[!grepl("^r\\d+", data$Alternative), ]
    local_pos <- data[data$Phi >= 0, ]
    local_neg <- data[data$Phi < 0, ]
    Local1_categories[[paste0("LocalSearch1 ", matrix_name, "(+)")]] <- local_pos
    Local1_categories[[paste0("LocalSearch1 ", matrix_name, "(-)")]] <- local_neg
    matrix_evaluation <- ME
  }
  # Step 4. Local search 2
  Local2_categories <- list()
  for (i in 1:(num_categories - 1)) {
    category_higher <- paste0("LocalSearch1 C", i, "(-)")
    category_lower <- paste0("LocalSearch1 C", i + 1, "(+)")
    matrix_higher <- Local1_categories[[category_higher]]
    matrix_lower <- Local1_categories[[category_lower]]
    Alt_higher <- matrix_higher$Alternative
    Alt_lower <- matrix_lower$Alternative
    Alt_filtered <- c(Alt_higher, Alt_lower)
    ME <- matrix_evaluation
    matrix_evaluation <- matrix_evaluation[matrix_evaluation$Alternative %in% Alt_filtered, ]
    # PROMETHEE
    RS <- PROMETHEEII(matrix_evaluation, data_criteria)
    data <- as.data.frame(RS[[1]])
    data$Phi <- as.numeric(data$Phi)
    local_pos <- data[data$Phi >= 0, ]
    local_neg <- data[data$Phi < 0, ]
    Local2_categories[[paste0("LocalSearch2 C", i, "- & C", i + 1, "+ (+)")]] <- local_pos
    Local2_categories[[paste0("LocalSearch2 C", i, "- & C", i + 1, "+ (-)")]] <- local_neg
    matrix_evaluation <- ME
  }
  # Step 5. Final classification
  Classifica <- data.frame(Alternative = matrix_evaluation$Alternative, Category = NA)
  # Classification: C1, Ck, and other groups.
  for (i in seq_along(Global_Search_categories)) {
    if (i == 1) {
      matrix_higher <- Local1_categories[["LocalSearch1 C1(+)"]]
      matrix_lower <- Local2_categories[["LocalSearch2 C1- & C2+ (+)"]]
      Alt_higher <- matrix_higher$Alternative
      Alt_lower <- matrix_lower$Alternative
      Alt_filtered <- c(Alt_higher, Alt_lower)
      Classifica$Category[Classifica$Alternative %in% Alt_filtered] <- i
    } else if (i == length(Global_Search_categories)) {
      matrix_higher <- Local1_categories[[paste0("LocalSearch1 C", i, "(-)")]]
      matrix_lower <- Local2_categories[[paste0("LocalSearch2 C", i - 1, "- & C", i, "+ (-)")]]
      Alt_higher <- matrix_higher$Alternative
      Alt_lower <- matrix_lower$Alternative
      Alt_filtered <- c(Alt_higher, Alt_lower)
      Classifica$Category[Classifica$Alternative %in% Alt_filtered] <- i
    } else {
      matrix_higher <- Local2_categories[[paste0("LocalSearch2 C", i - 1, "- & C", i, "+ (-)")]]
      matrix_lower <- Local2_categories[[paste0("LocalSearch2 C", i, "- & C", i + 1, "+ (+)")]]
      Alt_higher <- matrix_higher$Alternative
      Alt_lower <- matrix_lower$Alternative
      Alt_filtered <- c(Alt_higher, Alt_lower)
      Classifica$Category[Classifica$Alternative %in% Alt_filtered] <- i
    }
  }
  RS <- list(
    Global = Global_Search,
    Local1 = Local1_categories,
    Local2 = Local2_categories,
    Class = Classifica
  )
  return(RS)
}
